File: apps/vscode-extension/fixtures/script-flow/instrumented_sample.py
from instrumentation.scriptflow_trace import ScriptFlowTracer


tracer = ScriptFlowTracer(
    script_path=__file__,
    machine_id="local-dev",
)

# Copy these from the Script Flow node IDs for the final instrumented script.
NODE_SUMMARIZE = "sf1:fn:summarize_scores:l24c1"
NODE_LOOP = "sf1:loop:for-score-in-scores:l37c9"
NODE_EXCEPT = "sf1:except:except-valueerror:l42c13"


def normalize_score(score):
    if score < 0:
        raise ValueError("score must be positive")
    return score


def log_invalid(score):
    print(f"invalid score: {score}")


def summarize_scores(scores):
    with tracer.span(NODE_SUMMARIZE, span_id="span-summarize"):
        total = 0

        if not scores:
            return total

        loop = tracer.loop_aggregator(
            NODE_LOOP,
            span_id="span-loop",
            parent_span_id="span-summarize",
            flush_every_iterations=1000,
        )
        for score in scores:
            try:
                loop.measure(lambda: normalize_score(score))
                total += score
            except ValueError:
                with tracer.span(
                    NODE_EXCEPT,
                    span_id=f"span-invalid-{score}",
                    parent_span_id="span-summarize",
                ):
                    log_invalid(score)
        loop.flush("loop_exit")

        return total


if __name__ == "__main__":
    tracer.start_run({"invocation": "python instrumented_sample.py"})
    run_status = "ok"
    try:
        summarize_scores([1, 2, -1, 3])
    except Exception:
        run_status = "error"
        raise
    finally:
        tracer.end_run(run_status)

File: apps/vscode-extension/fixtures/script-flow/instrumented-sample.ts
import { fileURLToPath } from "node:url";

import { createScriptFlowTracer } from "../../src/scriptFlow/instrumentation/typescriptTrace.js";

const scriptPath = fileURLToPath(import.meta.url);
const tracer = createScriptFlowTracer({
  scriptPath,
  machineId: "local-dev",
  processId: process.pid,
});

// Copy these from the Script Flow node IDs for the final instrumented script.
const NODE_ACCUMULATE = "sf1:fn:accumulate:l12c1";
const NODE_GUARD = "sf1:branch:if-limit-0:l28c11";
const NODE_LOOP = "sf1:loop:for-index-limit:l45c7";
const ENTITY_ACCUMULATE = "apps/vscode-extension/fixtures/script-flow/instrumented-sample.ts#accumulate";

export async function accumulate(limit: number) {
  return tracer.span(
    {
      nodeId: NODE_ACCUMULATE,
      entityId: ENTITY_ACCUMULATE,
      spanId: "span-accumulate",
    },
    async () => {
      let total = 0;

      await tracer.span(
        {
          nodeId: NODE_GUARD,
          entityId: ENTITY_ACCUMULATE,
          spanId: "span-guard",
          parentSpanId: "span-accumulate",
        },
        async () => {
          if (limit <= 0) {
            total = 0;
          }
        },
      );

      if (limit <= 0) {
        return total;
      }

      const loop = tracer.createLoopAggregator({
        nodeId: NODE_LOOP,
        entityId: ENTITY_ACCUMULATE,
        spanId: "span-loop",
        parentSpanId: "span-accumulate",
        flushEveryIterations: 1000,
      });

      for (let index = 0; index < limit; index += 1) {
        loop.measure(() => {
          total += index;
        });
      }
      loop.flush("loop_exit");

      return total;
    },
  );
}

async function main() {
  tracer.startRun({ invocation: "tsx instrumented-sample.ts" });
  let status: "ok" | "error" = "ok";
  try {
    await accumulate(2500);
  } catch (error) {
    status = "error";
    throw error;
  } finally {
    tracer.endRun(status);
  }
}

if (process.argv[1] === scriptPath) {
  main().catch((error) => {
    console.error(error);
    process.exitCode = 1;
  });
}
